synapse/handlers/profile.py

```diff
@@ -89,6 +89,14 @@ async def get_profile(self, user_id):
             except RequestSendFailed as e:
                 raise SynapseError(502, "Failed to fetch profile") from e
             except HttpResponseException as e:
+                if e.code < 500 and e.code != 404:
+                    # Other codes are not allowed in c2s API
+                    logger.error(
+                        "Server replied with wrong response: %s %s", e.code, e.msg
+                    )
+                    # Change to 500 not to confuse clients
+                    e.code = 500
+                    e.msg = "Remote server replied: {} {}".format(e.code, e.msg)
```

Comment:
I think we should create a new error here, rather than fiddling with the exception received. Also, we may as well use the same exception as the case where we fail to connect to the remote above, which returns a 502.

So something like:

```suggestion
                if e.code < 500 and e.code != 404:
                    # Other codes are not allowed in c2s API
                    logger.info(
                        "Server replied with wrong response: %s %s", e.code, e.msg
                    )
                    
                    raise SynapseError(502, "Failed to fetch profile")
```

We don't want to log as an error either, as its not an error that the server admins care or do anything about.

Reply:
OK, that seems reasonable, commited.